### PR TITLE
#675 Fix className Composition

### PR DESCRIPTION
**Pull request by LucasUnplugged**
# Description

Addresses issue #675: `className` sticks to initial component, when composing Stitches-styled components.

**NOTE 1**: I renamed `initialClassName` to `latestClassName`, to reflect the change in line 316 of `css.js`, but the only functional change is the one to line 316.

**NOTE 2**: I don't have full context on why `initialClassName` was being used for chaining of class names; switching to `latestClassName` definitely addresses the issue above, but I'm not sure if it could have unintended consequences, due to my lack of context here.

### CODE REVIEW

**Comment by LucasUnplugged** [comment]
CodeSandbox showing the expected result: https://codesandbox.io/s/stitches-ci-cra-forked-5vkri
Versus CodeSandbox showing the issue: https://codesandbox.io/s/staging-field-4cs36

**Comment by peduarte** [comment]
Thanks for this! I tested the sandbox and looks good to me. I think it would be super useful to create a dedicated test for this use case though. I don't think we have a test specifically for this. 

About the code, I'll leave that for @jonathantneal

**Comment by LucasUnplugged** [comment]
> I think it would be super useful to create a dedicated test for this use case though. I don't think we have a test specifically for this. 

Yes, great point. I'm sure I can come up with some tests pretty easily.

**Comment by peduarte** [comment]
Nice one. Thanks for that. We'll review shortly

**Comment by jonathantneal** [comment]
Hey @LucasUnplugged, one issue with this change is that it moves the problem to the other side of composition.

See https://codesandbox.io/s/stitches-ci-cra-forked-y0u3u?file=/src/App.tsx

Now, the last class in a composition may return some unintended matches.

**Comment by LucasUnplugged** [comment]
> Hey @LucasUnplugged, one issue with this change is that it moves the problem to the other side of composition.
> 
> See https://codesandbox.io/s/stitches-ci-cra-forked-y0u3u?file=/src/App.tsx
> 
> Now, the last class in a composition may return some unintended matches.

@jonathantneal I think that's a bit of a separate issue, and maybe not even an issue (sort of). What I mean by that is:
1. The issue happens with the current version, but instead of a sibling exhibiting the issue, a parent (or multiple) can, [as seen here](https://codesandbox.io/s/little-forest-hx1b3?file=/src/App.tsx).
2. As seen in both that sandbox, and in [this one (forked from yours)](https://codesandbox.io/s/stitches-ci-cra-forked-bzd97?file=/src/App.tsx), the issue is a factor of how the hashes are generated, I believe: same styles === same hash.
3. I say _"kind of"_ not an issue because of the fact that it's probably an unavoidable fact of the purely hash-based class generation, but having identical styles on two nearby components is probably pretty rare.

`<ShamelessSelfPromotion>`
That said, I think that [what I'm proposing here](https://github.com/modulz/stitches/discussions/677) would actually fix this issue, by creating classes that are a combination of a custom string + the generated hash; so you could have `c-UnderlinedBlue-hrMtea` and `c-UnderlinedPink-hrMtea`, and the hashes being the same won't cause a conflict :)
`</ShamelessSelfPromotion>`

**Comment by peduarte** [comment]
@lucasmotta thanks for that! 🙏 

Jonathan and I had a long chat about this yesterday. The conclusion is that whichever way we go, the problem will still exist. And it depends on how users are creating + targeting their components (composition, descendant selector, etc)

> the issue is a factor of how the hashes are generated, I believe: same styles === same hash.

You're right. 

> I think that what I'm proposing here would actually fix this issue, by creating classes that are a combination of a custom string + the generated hash

Crating the hash from something more than just the style object would likely solve this issue. However, as I [explained in this discussion](https://github.com/modulz/stitches/discussions/677#discussioncomment-1086608), your current proposal isn't going to work as a third argument. 

I'm also concerned about the DX. As a user, I wouldn't want to have to give every element a friendly class name, especially when I am already naming the const. Feels like doing the same thing twice. This is problematic since it means the way to create more unique hashes would be to rely on babel, and this is something we don't want Stitches _to have to depend on_.

---

I think our options are:
- Leave it as is (issue exists in one end)
- Accept this PR (issue exists in the opposite end)
- Stop stacking classes (issue exists but since classes are not being stacked, it's a lot less likely to happen)

Thoughts?

**Comment by LucasUnplugged** [comment]
> Crating the hash from something more than just the style object would likely solve this issue. However, as I [explained in this discussion](https://github.com/modulz/stitches/discussions/677#discussioncomment-1086608), your current proposal isn't going to work as a third argument. 

Right, in vanilla JS I guess that's true; I didn't think about that. In TS, it throws an error if you try to repeat element params.

That said, that's easily fixable by targeting the last param, and checking if it's a string. I think that would cover all TS and JS use-cases.

> I'm also concerned about the DX. As a user, I wouldn't want to have to give every element a friendly class name, especially when I am already naming the const. Feels like doing the same thing twice. This is problematic since it means the way to create more unique hashes would be to rely on babel, and this is something we don't want Stitches _to have to depend on_.

Yeah, I completely understand this sentiment. I think that a solution like mine doesn't preclude that, if we keep it a "hidden feature" and promote the babel plugin as a solution.

The reason why I think we may want a component of this solution to be in Stitches itself, is because we can then make sure these custom classes work exactly as classes currently do in Stitches, including stacking, and having the styles associated with the styles. Anyhow, I'll continue that discussion there.

> I think our options are:
> 
> - Leave it as is (issue exists in one end)
> 
> - Accept this PR (issue exists in the opposite end)
> 
> - Stop stacking classes (issue exists but since classes are not being stacked, it's a lot less likely to happen)

I'm still in favour of No.2, because No.1 has more than just this secondary issue (hash-generation): it also has the primary issue of breaking use of composed components inside nested style objects. I think that issue can be much more pervasive, in codebases that use composition, than two nearby components having identical styles.

Remember: the hash issue is only likely an issue for siblings and children of siblings, because of encapsulation. I don't see a real need to use components as selectors with global styles: just put those styles in those components — and if you want to avoid repetition, create a separate object for such styles and spread them on the component definition.

So saying all that, I think No.3 is probably unnecessary. I think the reason the hash issue hasn't come up until now is because that's a much more rare problem than the composition one, which could happen a lot in some codebases.

(Sorry for the long comment 😅)

**Comment by LucasUnplugged** [comment]
BTW, if this gets approved, should I update the version numbers in my PR, or is that a whole separate process?

**Comment by peduarte** [comment]
@LucasUnplugged Thank you 🙌 

To be honest, right now I am really struggling to believe that there's a difference between option 1 and 2 (first or last class composition). 

I _think_ I agree that with option 2 it's less likely to happen. 

> So saying all that, I think No.3 is probably unnecessary.

Why do you think that?

---

We need to wait to hear Jonathan's thoughts and decide whether we can merge this or not. 

To be transparent with you, our current priority is to get Stitches stable from an API perspective. This is more like a bug fix so it's not currently right at the top of our list. But I'll see if we can get an answer ASAP.

You can see our roadmap here: https://github.com/modulz/stitches/projects/4

**Comment by LucasUnplugged** [comment]
Understood on the priority front.

For No.3, maybe I'm misunderstanding, but wouldn't that mean a whole revamp of how styles are associated with classes? If you have a stack of 3 classes on a component now, and want to reduce it to a single class, all those styles would have to get merged into that one class, right?

Maybe that's simpler to do than I'm imagining.

I'll try to create a sandbox to capture what I think is the difference between 1 and 2.

**Comment by LucasUnplugged** [comment]
@peduarte Okay, I've made two sandboxes to try to highlight the differences:

1. [This sandbox](https://codesandbox.io/s/nervous-surf-6ir27?file=/src/App.tsx) uses the existing version of Stitches. You can see that the footer is not fine in any of the 3 blocks, because it should never have a purple background. The footer incorrectly inherits that style from both the cases where composition is at work, and cases where identical styles === identical hashes.

2. [This other sandbox](https://codesandbox.io/s/stitches-ci-cra-forked-4lxqx?file=/src/App.tsx) uses my version from this PR. The footer is correct in any case where only composition is at work. The hash issue is still present in the second block, but it can be addressed via a variant, as seen in block 3 — and the variant code wasn't even repeated in that solution.

So in summary, I think conflicts due to same hash are much easier to avoid than conflicts due to inheritance, in codebases that make heavy use of (primitive) composition. I've already seen this in the wild. Passing the exact same styles to two nearby components seems rare, and likely fixable via variants instead.

That's my 2 cents, anyway :)

**Comment by jonathantneal** [comment]
In the current canary, we are using a `:where()` selector to target all of the elements in a composition without increasing specificity. It may resolve this issue, without choosing which side of the composition to use.

**Comment by LucasUnplugged** [comment]
> In the current canary, we are using a `:where()` selector to target all of the elements in a composition without increasing specificity. It may resolve this issue, without choosing which side of the composition to use.

@jonathantneal Oh, very interesting! I tried a recent sandbox with my example above, and it does seem to fix the issue (🎉); but I'm having trouble with nesting of styled components. Did the syntax for that changes? It seems to be working, by Typescript is screaming at me: https://codesandbox.io/s/objective-sound-57d99?file=/src/App.tsx

So before, I used to do something like this:
```TS
const Main = styled(Box, {
  background: '$gray500',
  [`& ${Header}`]: {
    background: '$purple600',
    color: 'white',
  },
});
```
(Where `Header` is a Stitches-styled component)

**Comment by peduarte** [comment]
Removing this from the V1 roadmap. We'll have to come back to this after launch, with more time, to fully understand the issue, side effects, proposed solutions, etc.

**Comment by LucasUnplugged** [comment]
> Removing this from the V1 roadmap. We'll have to come back to this after launch, with more time, to fully understand the issue, side effects, proposed solutions, etc.

Actually, v1 seems to have fixed this issue, @peduarte! And the type issue I last mentioned above also seems to have been fixed, since the earlier canary versions 🎉 [Sample sandbox](https://codesandbox.io/s/stitches-ci-cra-forked-zwdqr?file=/src/App.tsx) showing it working just fine.

As such, I'm closing this PR.

**Comment by peduarte** [comment]
Glad to hear :smi
